---
title: "`MADloy`: Detecting mosaic loss of Y events using SNP array data"
subtitle: "Marcos Lopez-Sanchez and Juan R. Gonzalez"
author: |
  | Institute for Global Health (ISGlobal), Barcelona, Spain
  | Bioinformatics Research Group in Epidemiolgy (BRGE)
  | (<http://www.creal.cat/brge.htm>)
date: "`r format(Sys.Date(), '%d %B %Y')`"
package: "`r paste('MADloy', packageVersion('MADloy'))`"
output:    
  BiocStyle::html_document:
    number_sections: true
    toc: yes
    fig_caption: yes
    fig_height: 4.5
    fig_width: 6
vignette: >
  %\VignetteIndexEntry{`MADloy`: Detecting mosaic loss of Y events (LOY) on SNP and NGS data}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---


# Introduction

`MADloy` is a package to detect mosaic loss of chromosome Y (LOY) events from NGS and SNP array data. The statistical method to detect LOY depends on the type of data we are dealing with since it relies on different distributional assumptions since NGS (e.g coverage) ans SNP array data (e.g log R ratio) have their own characteristics. 

In this vignettes we are going to illustrate how to process SNP array data. In that case, log R ratio (LRR) is used to stablished whether an individual is carrying an alteration in chromosome Y or not. We use Normal Inverse Gaussian (NIG) distribution which is able to model data that is obtained from the ratio of two values (e.g. LRR). The model parameters are estimated by using the trimmed mean LRR from the autosomes. Then, the values in the target region of chromosome Y are tested to provide evidences against the null hypothesis that corresponds to the case of having data from the reference. This figure illustrates how the NIG model perfectly fits the observed data, while the normal model is not.


```{r, fitNIG, echo=FALSE, message = FALSE, warning=FALSE, fig.cap="Empirical cumulative distribution function (ecdf) of the observed data and ecdf estimated using NIG and Normal models of reference LRR in PRAD data of TCGA project."}
library(MADloy)
load("C:/Juan/CREAL/GitHub/PRADdata/pradLOY.rda")
lrr <- MADloy:::getSummary(pradLOY)
plotNIG(lrr[,2], tit="PRAD data of TCGA")
``` 


By default, the target region used to detect LOY is the one in chromosome Y that avoids the PAR1 and PAR2 regions shared with chromosome X (mLRR-Y). In order to take into consideration the variability of coverage or LRR in a given sample, the method clusters individuals into LOY, normal and gain categories by using the NIG distribution evaluated at mLRR-Y region using parameters model estimated in the autosomes that are used as the reference values of LRR. The package summarizes the trimmed mean LRR in the probes contained in the SNP array and then perform the calling by using this pre-processed data. 

# Getting started

You can install `MADloy` from Github by typing 

```{r eval=FALSE}
devtools::install_github("isglobal-brge/MADloy")
```

Then the package is loaded as usual

```{r, message = FALSE}
library(MADloy)
```

# Processing SNP data with `madloy`: loading and summarizing LRR

`MADloy` contains a different function to pre-process your data depending on the technology. `madloy` processes SNP array data, while `madseqloy` deals with NGS data. Let us focus on `madloy`function (`madseqloy` and how to process NGS data is illustrated in a separate vignette). [**Marcos** no se si quitar lo de madseqloy de aquí para no dar pistas o para que nos pidan que lo pongamos]. 

Detecting LOY events in SNP arrays requires to get summarized data from log R ratio (LRR) files. The idea is to see whether LRR in the male-specific region of chromosome Y is different from 0. The mLRR-Y corresponds to the 56-Mb region between pseudoautosomal regions 1 and 2 (PAR1 and PAR2) on chromosome Y (chrY:2,694,521-59,034,049, hg19/GRCh37). 

If your data has been obtained from Illumina arrays, LRR can be obtained from the reports generated with BeadStudio tool (see for instance `MAD` package documentation [here](http://www.creal.cat/media/upload/arxius/jr/GADA/mad_vignette.pdf)), or other R packages as [`crlmm`](http://bioconductor.org/packages/release/bioc/html/crlmm.html) or [`beadarraySNP`](https://www.bioconductor.org/packages/release/bioc/html/beadarraySNP.html). If your data has been generated from any Affymetrix chip (e.g. GenomeWide, Axiom or CytoScan) you can use the either Affymetrix Power Tools (APT) or `affy2sv` package [`affy2sv`](http://bioconductor.org/packages/release/bioc/html/affy2sv.html). The function to process LRR files only requires the path indicanting were those files are located. Let us illustrate how to get summarized data from 3 samples having information about LRR that are available at `MADloy` package. The folder can be get by writting:


```{r, MADloy_1}
rawDataPath <- system.file("extdata", "rawData", package = "MADloy")
rawDataPath
dir(rawDataPath)
```
Then, data can be summarized by usgin  `madloy` function. The function creates an object of class `MADloy` that can be inspected by using the generic `print` function.
```{r, MADloy_1_b}
res <- madloy(rawDataPath)
res
```
We can manually inspect  information of mLRR-Y region (e.g., one by one sample) and decide whether a given individulas is having a LOY or not. For example this code will create the plot of Sample1 where LRR in the mLRR-Y region (shaded area) is clearly lost.

```{r, plotSample1,  fig.cap="LRR in the mLRR-Y region (shaded) of Sample1 from the illustrative example"}
plotIndSNP(res, sample="Sample1")
````

The same figure can be obtained of Sample2 which is not having a LOY event:

```{r, plotSample2,  fig.cap="LRR in the mLRR.Y region (shaded) of Sample2 from the illustrative example"}
plotIndSNP(res, sample="Sample2")
````

However, when thousands of samples are analyzed a proper statistical method must be used to differenciate those samples that are having LOY. Next section illustrate how to visualize the information of mLRR-Y of lot of samples


# Visualizing a set of samples
Let us illustrate how to visualize the data using 89 CEU HapMap samples.  Already pre-processed data can be retrieved into your ``R`` session from the `MADloyHapMapAll` data set that is included in the package by simply typing: 

```{r, MADloy_3}
data(MADloy_HapMapAll)
MADloyHapMapAll
```
Data can be visually inspected by using the generic `plot` function. This function depicts the mean difference between the LLR of Y chromosome and the reference region that is performed in order to control for possible technical artifacts. As previously mention, the reference is consider the autosomes and LRR is summarized by using the trimmed mean in order to remove the effect of having any gain or lose in the genome. Notice that this reference chromosome can be changed by the user. 

```{r, fig.show='hold',  fig.cap = "Plot of MADloy object including 89 HapMap samples"}
plot(MADloyHapMapAll)
```

This figure shows several samples having LOY (those in the -2, -4 range of Y-axis) that correspond to female samples. Of course, these are not real LOY events since females are expected to show this pattern. Let us pay attention in the male individuals that are also available as a data example in our package. We have selected 23 male samples whose pre-processed LLR that can be get into ``R``by:

```{r, MADloy_4}
data(MADloy_HapMapMale)
MADloyHapMapMale
```
The plot is obtained by

```{r, fig.show='hold',  fig.cap = "Plot of a MADloy object with 23 HapMap 3 male samples analyzed"}
plot(MADloyHapMapMale, ylim=c(-0.2, 0.2))
```
We observe that all the points (that represent the mean value of LLR in mLRR-Y region of a given sample) are around 0. This indicates that all the samples can be considerd normal individuals with regard to their values in chromosome Y. 


# Detecting LOY events

Once samples are processed, an object of class `MADloy` is created. After that, one may want to detect those samples having an altered value in chromosome Y. These alteration can be either a LOY event or a gain. The calling is performed by using the function `getLOY`. This function uses Normal Inverse Gaussian (NIG) distribution to model LRR. Model parameters are estimated by using trimmed mean LRR in the autosomes that serve as the reference. Trimmed mean aims to remove those values in autosomal regions having copy number alterations. Then, the probability that LRR in the mLRR-Y region belongs to the reference distribution is assessed by using the cummulative NIG distribution. The main manuscript furter describes how these methods are implemented.

Let us start by illustrating how to detect LOY events in samples from TCGA. We have donwladed .... (** Marcos describir **) . Once .txt files have been created, the data is processed by using `madloy` function as previously described:

```{r, process_BLCA, eval=FALSE}
BLCA <- madloy("c:/Juan/CREAL/GitHub/BLCAdata/rawData/")
``` 

This processed object can be loaded from our `BRGEdata` package. The package can be installed from Bioconductor.

```{r, get_BLCA, eval=FALSE}
library(BRGEdata)
data(BLCA)
```

**To be removed: ** So far it can be loaded from
```{r, getBLCA_no}
load("C:/Juan/CREAL/GitHub/BLCAdata/blcaLOY.rda")
``` 

First, let's have a look at the summarized data of LRR in mLRR-Y region and in the autosomal choromosomes that are used as the reference ones

```{r, lrr_ref_mLRR-Y}
lrr.summary <- MADloy:::getSummary(blcaLOY)
head(lrr.summary)
```

```{r, plot_lrr_ref_mLRR-Y, fig.show='hold', fig.height=9, fig.cap ="LRR in reference (autosomes) and mLRR-Y region of BLCA data"}
par(mfrow=c(2,1))
plot(lrr.summary[,2], xlab="samples", ylab="LRR in autosomes")
abline(h=0, lty=2, col="red")
plot(lrr.summary[,1], xlab="samples", ylab="LRR in mLRR-Y region")
abline(h=0, lty=2, col="red")
```

Here we can observe that:

> * The LRR in autosomes are centered at 0 as expected
> * The LRR in mLRR-Y region deviates from 0 since there is only one copy (e.g ploidy is equal to 1) and hence, the LRR is centered in 2/3·log(1/2) = `r round(2/3*log(1/2),2)`
> * There is also a little deviation due to the overall mean in the array

Therefore, data in mLRR-Y region is normalized by computing 

```{r, normalized}
norm.lrr <- lrr.summary[,1] - lrr.summary[,2] - 2/3*log(1/2)
``` 

Now the normalized LRR in mLRR-Y regions is properly centered at 0

```{r, plot_normalized, fig.show='hold', fig.width=6, fig.height=5, fig.cap ="Normalized LRR in mLRR-Y region of BLCA data"}
plot(norm.lrr, ylab="Normalized LRR in mLRR-Y region", xlab="samples")
abline(h=0, lty=2, col="red")
``` 

The calling that also incorporates the normalization procedure can be performed by computing:

```{r, BLCA_analysis}
n <- length(blcaLOY$par$files)
call.blca <- getLOY(blcaLOY, offset= 0.46, pval.sig=0.05/n)
call.blca
```

Notice that the offset corresponds to the absolute value of the mean LRR of chromosome Y in this array. In some occasions the software used to obtain LRR normalized this value to 0. In that case the offset is not necesary. The argument `pval.sig` controls the FDR of alterations (LOY or gains) by using Bonferroni correction.

And the figure to visually inspect the alterations in Y chormosoem are be obtained by using the generic `plot` function: 

```{r, fig.show='hold', fig.cap = "Plot of a LOY object obtained from BLCA dataset"}
plot(call.blca, pos.leg="topright")
```




The plot shows samples having a LOY event as well as gains. This can be verified by having a look at the LRR data in chromosome Y. We have incorporated different text files into the package having the LRR information of samples `TCGA-CF-A3MI-01A` and `TCGA-E7-A6ME-01A`

```{r, show_NAsamples}
PathBlcaLRR <- system.file("extdata", package = "MADloy")
list.files(PathBlcaLRR, pattern=".txt")
```

The plot of LRR in mLRR-Y region of the sample `TCGA-CF-A3MI-01A` indicates that this is a real gain

```{r, plot_gain, fig.show='hold', fig.cap = "Plot of a LOY object obtained from SNP array data"}
plotIndSNP(file.path(PathBlcaLRR, "TCGA-CF-A3MI-01A.txt"))
```

On the other hand, the same plot of the sample `TCGA-E7-A6ME-01A` describes a LOY event

```{r, plot_loy, fig.show='hold', fig.cap = "Plot of a LOY object obtained from SNP array data"}
plotIndSNP(file.path(PathBlcaLRR, "TCGA-E7-A6ME-01A.txt"))
```

Notice that the function `plotIndSNP` can be used to visualize the mLRR-Y region by passing different inputs. In that case, we have provided the path and the name of the file containing the LRR data. Another option is to provide and object of class `MADloy` or a `data.frame` containing the rs, chromosome, position and LRR information.