---
title: "`MADloy`: Detecting mosaic loss of Y events using SNP array data"
subtitle: "Marcos Lopez-Sanchez, Juan R. Gonzalez"
author: |
  | Institute for Global Health (ISGlobal), Barcelona, Spain
  | Bioinformatics Research Group in Epidemiolgy (BRGE)
  | (<http://www.creal.cat/brge.htm>)
date: "`r format(Sys.Date(), '%d %B %Y')`"
package: "`r paste('MADloy', packageVersion('MADloy'))`"
output:    
  BiocStyle::html_document:
    number_sections: true
    toc: yes
    fig_caption: yes
    fig_height: 3
    fig_width: 4
vignette: >
  %\VignetteIndexEntry{`MADloy`: Detecting mosaic loss of Y events (LOY) on SNP and NGS data}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---


# Introduction

`MADloy` is a package to detect mosaic loss of chromosome Y (LOY) events from NGS and SNP array data. In this vignettes we are going to illustrate how to process SNP array data. The statistical method to detect LOY relies on different distributions since each type of data are having their own characteristics. For the case of SNP array data, log R ratio (LRR) is used to stablished whether an individual is carrying this alteration or not. We use Normal Inverse Gaussian (NIG) distribution which is able to model data that is obtained from the ratio of two values (e.g. LRR). The model parameters are estimated by using the LRR distribution in a reference chromosome. Then, the values in a target region of chromosome Y are tested to decipher whether the observed value belongs to the reference distribution (e.g. normal individuals) or not. 

By default, the region used to detect LOY is a region in chromosome Y region that avoids the PAR1 and PAR2 regions shared with chromosome X (mY). In order to take into consideration the variability of coverage or LRR in a given sample, the method clusters individuals into LOY, normal and GAIN categories by using a NIG distribution evaluated at mLRR-Y region and whole chromosome 22 which is used as the reference one. Nonetheless, the reference region can be set manually. The package summarizes the mean LRR in the probes contained in the SNP array and then perform the calling by using this pre-processed data. 

You can install `MADloy` from Github by typing 

```{r eval=FALSE}
devtools::install_github("isglobal-brge/MADloy")
```

```{r, message = FALSE}
library(MADloy)
```

# Processing SNP data with `madloy` (load and summarize)

`MADloy` contains a different function to pre-process your data depending on the technology. `madloy` processes SNP array data, while `madseqloy` deals with Exome Sequencing data. Next, we illustrate how to preform  this step when having SNP arrays.

Detecting LOY events in SNP arrays requires to get summarized data from log R ratio (LRR) files. The idea is to see whether LRR in the male-specific region of chromosome Y (mLRR-Y) is different from 0. The mLRR-Y corresponds to the 56-Mb region between pseudoautosomal regions 1 and 2 (PAR1 and PAR2) on chromosome Y (chrY:2,694,521-59,034,049, hg19/GRCh37). 

If your data has been obtained from Illumina arrays, LRR can be obtained from the reports generated with BeadStudio tool (see for instance `MAD` package documentation [here](http://www.creal.cat/media/upload/arxius/jr/GADA/mad_vignette.pdf)), or other R packages as [`crlmm`](http://bioconductor.org/packages/release/bioc/html/crlmm.html) or [`beadarraySNP`](https://www.bioconductor.org/packages/release/bioc/html/beadarraySNP.html). If your data has been generated from any Affymetrix chip (e.g. GenomeWide, Axiom or CytoScan) you can use the either Affymetrix Power Tools (APT) or `affy2sv` package [`affy2sv`](http://bioconductor.org/packages/release/bioc/html/affy2sv.html). The function to process LRR files only requires the path indicanting were those files are located. Let us illustrate how to get summarized data from 3 samples having information about LRR that are available at `MADloy` package. The folder can be get by writting:


```{r, MADloy_1}
rawDataPath <- system.file("extdata", "rawData", package = "MADloy")
rawDataPath
dir(rawDataPath)
```
Then, data can be summarized by usgin  `madloy` function. The function creates an object of class `MADloy` that can be inspected by using the generic `print` function.
```{r, MADloy_1_b}
res <- madloy(rawDataPath)
res
```
We can manually inspect  information of mLRR-Y region (e.g., one by one sample) and decide whether a given individulas is having a LOY or not. For example this code will create the plot of Sample1 where LRR in the mLRR-Y region (shaded area) is clearly lost.

```{r, plotSample1, fig.width=6, fig.height=4.5, fig.cap="LRR in the mLRR.Y region (shaded) of Sample1 from the illustrative example"}
plotIndSNP(res, sample="Sample1")
````

The same figure can be obtained of Sample2 which is not having a LOY event:

```{r, plotSample2, fig.width=6, fig.height=4.5, fig.cap="LRR in the mLRR.Y region (shaded) of Sample2 from the illustrative example"}
plotIndSNP(res, sample="Sample2")
````

However, when thousands of samples are analyzed a proper statistical method can be used to differenciate those samples that are having LOY. Next section illustrate how to visualize the information of mLRR-Y of lot of samples


# Visualizing a set of SNP samples
Let us illustrate how to visualize the data using 89 CEU HapMap samples.  Already pre-processed data can be retrieved into your ``R`` session from the `MADloyHapMapAll` data set that is included in the package by simply typing: 

```{r, MADloy_3}
data(MADloy_HapMapAll)
MADloyHapMapAll
```
Data can be visually inspected by using the generic `plot` function. This function depicts the median LLR of Y chormosome adjusted by the LRR of chromosome 22 in order to control for possible technical artifacts. Notice that this reference chromosome can be changed if the user suspects that alterations in this chromosome can appear in the analyzed data. 

```{r, fig.show='hold', fig.width=8, fig.height=6, fig.cap = "Plot of MADloy object including 89 HapMap samples"}
plot(MADloyHapMapAll)
```

This figure shows several samples having LOY (those in the -2, -4 range of Y-axis) that correspond to female samples. Of course, these are not real LOY samples since females are expected to show this pattern. Let us pay attention in the male individuals that are also available as a data example in our package. We have selected 23 male samples whose pre-processed LLR that can be get into ``R``by:

```{r, MADloy_4}
data(MADloy_HapMapMale)
MADloyHapMapMale
```
The plot is obtained by

```{r, fig.show='hold', fig.width=8, fig.height=6, fig.cap = "Plot of a MADloy object with 23 HapMap 3 male samples analyzed"}
plot(MADloyHapMapMale, ylim=c(-0.3, 0.3))
```
We observe that there are some points that deviate from 0. The question is to determine whether those samples belong to the general population of normal samples.



# Detecting LOY events

Once samples have been processed, the user would have created a `MADloy` object. The next step is to detect those samples having a LOY event. The calling is performed by using the function `getLOY`. This function uses the NIG distribution. The main manuscript describes how these methods are implemented to detect LOY events in both type of data.

Let us start by illustrating how to detect LOY in the HapMap samples. The object containing preprocessed data is called `MADloyHapMapMale`. Calling is performed by typing:

```{r, getLOY}
nsamples <- length(MADloyHapMapMale$par$files)
ans <- getLOY(MADloyHapMapMale, pval.sig = 0.05/nsamples)

```

Results can be visualized with the generic `plot` function as follows:

```{r, fig.show='hold', fig.width=6, fig.height=5, fig.cap = "Plot of a LOY object obtained from SNP array data"}
plot(ans)
```

The method is indicating that there are three samples having a LOY that can be verified by looking at the raw data. We have incorporated a text file into the package having the LRR information of samples NA1276, NA12144 and NA11829 who are having a LOY and NA07048 who is not.

```{r, show_NAsamples}
PathHapMapLRR <- system.file("extdata", package = "MADloy")
list.files(PathHapMapLRR, pattern=".txt")
```

```{r, fig.show='hold', fig.width=6, fig.height=5, fig.cap = "Plot of a LOY object obtained from SNP array data"}
plotIndSNP(file.path(PathHapMapLRR, "NA12716.txt"))
```


Let us illustrate how to perform this procedure using data from TCGA ..... Once .txt files have been created, the data is processed by executing:

```{r, process_BLCA, eval=FALSE}
BLCA <- madloy("c:/Juan/CREAL/GitHub/BLCAdata/rawData/")
``` 

This processed object can be loaded from our `BRGEdata` package. The package can be installed from Bioconductor.

```{r, get_BLCA, eval=FALSE}
library(BRGEdata)
data(BLCA)
```

**To be removed: ** So far it can be loaded from
```{r, getBLCA_no}
load("C:/Juan/CREAL/GitHub/BLCAdata/blcaLOY.rda")
``` 

```{r, BLCA_analysis}
call.blca <- getLOY(blcaLOY, offset= 0.46, pval.sig=0.05/49)
call.blca
```

And the figure to determine LOY samples can be obtained by: 

```{r, fig.show='hold', fig.width=6, fig.height=5,fig.cap = "Plot of a LOY object obtained from BLCA dataset"}
plot(call.blca)
```
